File: joss/comparisons/cat-1-column-schema-reference/15-distinct-narrowing/example-distinct-narrowing.R
# Error Class 15: Distinct Column Narrowing
#
# R's distinct() with .keep_all = FALSE drops non-specified columns,
# but with .keep_all = TRUE keeps all columns with arbitrary values.
# There is no type-level tracking of which columns remain.

library(tidyverse)

encounters <- tibble(
  patient_id = c("P001", "P001", "P002", "P002"),
  department = c("Cardiology", "Cardiology", "Emergency", "Primary Care"),
  encounter_type = c("Outpatient", "Inpatient", "ED", "Outpatient"),
  physician = c("Dr. Patel", "Dr. Patel", "Dr. Lee", "Dr. Martinez")
)

# distinct without .keep_all drops extra columns
unique_depts <- encounters %>% distinct(patient_id, department)
print(unique_depts)
# physician is gone — but no type-level indication

# distinct with .keep_all keeps all columns with arbitrary values
unique_depts_all <- encounters %>% distinct(patient_id, department, .keep_all = TRUE)
print(unique_depts_all)
# physician column present — which value is kept is arbitrary

File: joss/comparisons/cat-1-column-schema-reference/15-distinct-narrowing/example-distinct-narrowing.py
"""
Error Class 15: Distinct Column Narrowing

Python's drop_duplicates() keeps all columns — it does not narrow
the schema. You can still access any column after dedup, even though
the semantics may be ambiguous (which row's value is kept?).
"""
import pandas as pd

encounters = pd.DataFrame({
    "patient_id": ["P001", "P001", "P002", "P002"],
    "department": ["Cardiology", "Cardiology", "Emergency", "Primary Care"],
    "encounter_type": ["Outpatient", "Inpatient", "ED", "Outpatient"],
    "physician": ["Dr. Patel", "Dr. Patel", "Dr. Lee", "Dr. Martinez"],
})

# drop_duplicates keeps all columns — which physician is kept is arbitrary
unique_depts = encounters.drop_duplicates(subset=["patient_id", "department"])
print(unique_depts)
# physician column still present — but which value is undefined behavior
# (depends on which row pandas keeps, usually first)

# No warning that physician value is arbitrary for the kept rows
print(unique_depts["physician"].tolist())
